node/workflow-api.md

```diff
@@ -0,0 +1,322 @@
+# 3rd revision of the Typescript SDK Workflow API
+
+### Current API
+
+```ts
+import { Trigger } from '@temporalio/workflow';
+
+export type MyWorkflow = (
+  arg1: number,
+  arg2: string
+) => {
+  execute(): Promise<void>;
+  signals: {
+    unblock(): void;
+  };
+  queries: {
+    isBlocked(): boolean;
+  };
+};
+
+export const myWorkflow: MyWorkflow = (arg1, arg2) => {
+  const unblocked = new Trigger<void>();
+  let isBlocked = true;
+  return {
+    // Main workflow logic
+    async execute() {
+      await unblocked;
+      isBlocked = false;
+    },
+    signals: {
+      unblock() {
+        unblocked.resolve();
+      },
+    },
+    queries: {
+      isBlocked() {
+        return isBlocked;
+      },
+    },
+  };
+};
+```
+
+#### Issues
+
+- Types must be explicitly declared (To make [eslint happy](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/explicit-module-boundary-types.md))
+- Very verbose
+- Trigger (ResolvablePromise) is a foreign concept to JS developers
+- Trigger doesn't provide all of the functionality for this workflow, an additional `isBlocked` boolean variable is required
+- Signals and queries must be defined at the top level in order for types to be inferred.
+
+#### Advantages
+
+- Types for all Workflow handlers can be inferred at Workflow handle creation
+
+### Goals for the redesign
+
+- Maintain type inference capabilities - Mandatory for the Typescript SDK
+- Dynamic signals and queries registration first - A single API for Workflow handler registration
+- Componentization of related functionality - increase reusability
+- Implicit interface declarations - less boilerplate
+- No global variables - to support running Workflows in less isolated runtimes
+
+### New proposal
+
+#### Remove the `Trigger` API in favor of `condition`
+
+`condition` is similar to `Workflow.await` in Java, it returns a Promise and accepts a closure which resolve the Promise when it returns `true`.
+
+```ts
+import { condition } from '@temporalio/workflow';
+
+let isBlocked = true;
+await condition(() => !isBlocked);
+```
+
+#### Remove the Workflow factory wrapper and always use dynamic handler registration
+
+`workflows.ts`
+
+```ts
+import { Signal, Query, condition } from '@temporalio/workflow';
+
+// Define the signal / query with the type signature at the top level to share the definition for handle creation
+export const unblockSignal = new Signal<[] /* args */>('unblock');
+export const isBlockedQuery = new Query<[] /* args */, boolean /* return type */>('isBlocked');
+
+// Types are inferred without an additional type definition
+export async function myWorkflow(arg1: number, arg2: string): Promise<void> {
+  let isBlocked = true;
+  unblockSignal.handler = () => (isBlocked = false);
+  isBlockedQuery.handler = () => isBlocked;
```

Comment:
May want to allow/note that handlers can support `Promise` returns (i.e. async callbacks)

Reply:
Only signal handlers can return promises, queries cannot.
I'm not sure if we should let condition return a promise too, seems like there's no need for that.

Reply:
>  I'm not sure if we should let condition return a promise too, seems like there's no need for that.

I suppose not since a `condition` can be implemented as a promise, e.g. (untested):

```typescript
let unblockResolve;
let unblock = new Promise(resolve => { unblockResolve = resolve })
unblockSignal.handler = unblockResolve
await unblock
```

^ Is that a reasonable alternative for how someone might implement a single unblock (i.e. not needing query or ability to re-block)?

Reply:
It's sort of like the trigger API we have now https://nodejs.temporal.io/api/classes/workflow.Trigger